File: java/DistinctSubsequences.java
public class DistinctSubsequences {
    public int numDistinct(String S, String T) {
        if (S == null || T == null) {
            return 0;
        }
        int n = S.length();
        int m = T.length();
        int[] dp = new int[m + 1];
        dp[0] = 1;
        for (int i = 1; i <= n; ++i) {
            for (int j = m; j >= 1; --j) {
                dp[j] += S.charAt(i - 1) == T.charAt(j - 1) ? dp[j - 1] : 0;
            }
        }
        return dp[m];
    }
}

File: python/distinct_subsequences.py
#!/usr/bin/env python3

class Solution:
    # @param two strings
    # @return an integer
    def numDistinct(self, S, T):
        n = len(S)
        m = len(T)
        if n < m:
            return 0
        dp = [0] * (m + 1)
        dp[m] = 1
        for i in reversed(range(n)):
            for j in range(m - min(m, n - i), m):
                if S[i] == T[j]:
                    dp[j] += dp[j + 1]
        return dp[0]

def main():
    solver = Solution()
    print(solver.numDistinct('rabbbit', 'rabbit'))

if __name__ == '__main__':
    main()
